### PR TITLE
More Optional Class Cleanup

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This PR removes the public API for encoding/decoding classes from all the languages that still have it.
And changes our logic to throw an exception if we try to skip an optional class.
Note that it is still possible for users to send/receive optional classes if they use the lower-level stream API's, but we can only do so much here. A sufficiently motivated user will always find a way to send these.

It also brings all the tests back into alignment with one another, and removes the last place where we were manually sending/checking for optional classes using the stream APIs.

----

The error messages that were added to C++ and C# were different, so I standardized on `cannot skip optional class`.
This seemed the most consistent with our other error messages in the marshaling/unmarshaling logic:
`cannot deserialize object`, `cannot unmarshal a proxy without a communicator`, `invalid Object slice`, `empty indirection table`, etc. Feel free to comment alternatives.

### CODE REVIEW

**Review by externl — APPROVED** [review]

File: matlab/lib/+Ice/InputStream.m [hunk 1]
```diff
@@ -635,7 +635,7 @@ function skipOptional(obj, format)
                     end
                     obj.skip(sz);
                 case Ice.OptionalFormat.Class
-                    obj.readValue([], '');
+                    throw(Ice.MarshalException('', '', sprintf('cannot skip an optional class')));
```
Comment (pepone) [inline matlab/lib/+Ice/InputStream.m:638]: Seems like `sprintf` is not required here, as we don't have format params

```suggestion
                    throw(Ice.MarshalException('', '', 'cannot skip an optional class'));
```

**Review by pepone — APPROVED** [review]

File: matlab/lib/+Ice/InputStream.m (hunk 1 above)
Comment (InsertCreativityHere) [reply matlab/lib/+Ice/InputStream.m]: Yep, good catch!

File: java/src/Ice/src/main/java/com/zeroc/Ice/InputStream.java [hunk 2]
```diff
@@ -1622,43 +1622,6 @@ public void readValue(java.util.function.Consumer<Value> cb) {
     readValue(cb, Value.class);
   }
```
Comment (bernardnormier) [inline java/src/Ice/src/main/java/com/zeroc/Ice/InputStream.java:1623]: Please simplify the readValue above now that (presumably) the callback is never null.

File: csharp/src/Ice/InputStream.cs [hunk 3]
```diff
@@ -2595,7 +2595,7 @@ private void skipOptional(OptionalFormat format)
             }
             case OptionalFormat.Class:
             {
-                throw new MarshalException("cannot unmarshal or skip an \"optional class\" parameter or field");
+                throw new MarshalException("cannot skip an optional class");
```
Comment (bernardnormier) [inline csharp/src/Ice/InputStream.cs:2598]: Please simplify readValue above. The callback is non-nullable now.

File: js/src/Ice/Stream.js [hunk 4]
```diff
@@ -1830,18 +1830,6 @@ class InputStream
             });
     }
 
-    readOptionalValue(tag, cb, T)
```
Comment (bernardnormier) [inline js/src/Ice/Stream.js:1833]: It's likely the cb in the readValue above can no longer be null.

File: swift/src/Ice/InputStream.swift [hunk 5]
```diff
@@ -239,7 +239,7 @@ public class InputStream {
         case .FSize:
             try skip(read())
         case .Class:
-            try read(UnknownSlicedValue.self, cb: nil)
+            throw MarshalException(reason: "cannot skip an optional class")
```
Comment (bernardnormier) [inline swift/src/Ice/InputStream.swift:242]: Ditto - no more nil cb.

File: matlab/lib/+IceInternal/EncapsDecoder10.m [hunk 6]
```diff
@@ -29,7 +29,7 @@ function readPendingValues(obj)
         end
 
         function readValue(obj, cb)
-            %assert(~isempty(cb));
```
[the remark this answers is not in the document]
Comment (bernardnormier) [reply matlab/lib/+IceInternal/EncapsDecoder10.m:32]: See other comments.

**Review by bernardnormier — APPROVED** [review]
Looks good.

I believe you forgot to update the InputStream C# code to remove nullable.

**Comment by InsertCreativityHere** [comment]
> I believe you forgot to update the InputStream C# code to remove nullable.

Which function are you talking about specifically?

For the main function that I touched: `internal override void readValue(System.Action<Value?> cb)`
I believe this is correct. The callback itself isn't nullable, but the value it receives is nullable because the decoding may of failed.

**Comment by bernardnormier** [comment]
> For the main function that I touched: `internal override void readValue(System.Action<Value?> cb)` I believe this is correct. The callback itself isn't nullable, but the value it receives is nullable because the decoding may of failed.

My mistake, the code is correct.

**Pull request merged**